terminate envoy when number of active connections is zero

ramaraochavali opened the pull request:
For https://github.com/istio/istio/issues/34855 except that it still uses terminationDrainDuration.

It gets `downstream_cx_active` of all listeners (ignores admin and stats listeners) and when the total becomes zero, it terminates Envoy or waits for terminationDrainDuration. It starts this process after waiting for minDrainDuration i.e. 5s.

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

istio-testing commented:
Skipping CI for Draft Pull Request.
If you want CI signal for your change, please convert it to an actual PR.
You can still manually trigger a test run with `/test all`

ramaraochavali commented:
@howardjohn @kyessenov I was experimenting this a bit and here is draft PR. Please take a look - if the approach makes sense - I will add tests.

ramaraochavali commented:
/test all

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435145974010875904) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435145974010875904) | `/test release-notes_istio`
lint_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435145972345737216) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435145974010875904) | `/test release-notes_istio`
lint_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435145972345737216) | `/test lint_istio`
gencheck_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/gencheck_istio/1435145973172015104) | `/test gencheck_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435145974010875904) | `/test release-notes_istio`
lint_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435145972345737216) | `/test lint_istio`
gencheck_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/gencheck_istio/1435145973172015104) | `/test gencheck_istio`
unit-tests_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435145966519848960) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/retest

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435145972345737216) | `/test lint_istio`
gencheck_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/gencheck_istio/1435145973172015104) | `/test gencheck_istio`
unit-tests_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435145966519848960) | `/test unit-tests_istio`
release-notes_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435157843559845888) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/gencheck_istio/1435145973172015104) | `/test gencheck_istio`
unit-tests_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435145966519848960) | `/test unit-tests_istio`
release-notes_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435157843559845888) | `/test release-notes_istio`
lint_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435157841877929984) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
gencheck_istio | ec4768491b84d6a65720e444fcd778d66e290133 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/gencheck_istio/1435145973172015104) | `/test gencheck_istio`
release-notes_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435157843559845888) | `/test release-notes_istio`
lint_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435157841877929984) | `/test lint_istio`
unit-tests_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435157836668604416) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test all

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435157841877929984) | `/test lint_istio`
unit-tests_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435157836668604416) | `/test unit-tests_istio`
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 371be4cefc1bca726668b99e7284c35e9e9797a9 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/lint_istio/1435157841877929984) | `/test lint_istio`
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435167440915927040) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | 9524530b5532178ab71134c3688b9f675da53b6f | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435174651931136000) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

ramaraochavali commented:
/test unit-tests_istio

ramaraochavali commented:
/test unit-tests_istio

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | 40bb7043ffb19ca67a50036b270a1c36560a00fc | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435200974334464000) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | 325986ce19a4e611c094d944c266fa1ee51a7d21 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435204116753158144) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

ramaraochavali commented:
/test unit-tests_istio

howardjohn commented on pkg/config/mesh/mesh.go:
```diff
@@ -43,7 +43,7 @@ func DefaultProxyConfig() meshconfig.ProxyConfig {
 		DrainDuration:            types.DurationProto(45 * time.Second),
 		ParentShutdownDuration:   types.DurationProto(60 * time.Second),
-		TerminationDrainDuration: types.DurationProto(5 * time.Second),
+		TerminationDrainDuration: types.DurationProto(45 * time.Second),
```
Discussed a bit more in the issue, but we need to make sure we don't make this a breaking API change

howardjohn commented on pkg/envoy/agent.go:
```diff
@@ -17,22 +17,43 @@ package envoy
+// minDrainDuration is the minimum duration for which agent waits before it actively checks
+// for outstanding connections. After this period, agent terminates proxy whenever active
+// connections become zero.
+var minDrainDuration = 5 * time.Second
```
same comment about breaking change. You should be able to override to 0s

howardjohn commented on pkg/envoy/agent.go:
```diff
@@ -17,22 +17,43 @@ package envoy
+var activeConnectionCheckDelay = 1 * time.Second
+
+var knownIstioListeners = sets.NewSet(
+	"listener.0.0.0.0_15006.downstream_cx_active", "listener.0.0.0.0_15021.downstream_cx_active",
```
The ports can be configured

howardjohn commented on pkg/envoy/agent.go:
```diff
@@ -105,13 +131,72 @@ func (a *Agent) terminate() {
-	time.Sleep(a.terminationDrainDuration)
-	log.Infof("Graceful termination period complete, terminating remaining proxies.")
-	a.abortCh <- errAbort
+	time.Sleep(minDrainDuration)
```
on L129 log what we are going to do maybe. Like `Agent draining proxy for 5s, then waiting for active connections to terminate...`

howardjohn commented on pkg/envoy/agent.go:
```diff
@@ -17,22 +17,43 @@ package envoy
 	return &Agent{
 		proxy:                    proxy,
 		statusCh:                 make(chan exitStatus, 1), // context might stop drainage
+		drainCh:                  make(chan bool),
```
```suggestion
		drainCh:                  make(chan struct{}),
```

then use `close(drainCh)` to signal, not send

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | f631a8bdc9527df6245ea11f0a731d246628d257 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435472798242836480) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kyessenov commented:
The cleanest solution is to wait for other containers to complete IMHO (and dealing with their organic restarts). 

What you have here seems reasonable. The danger is some app might want to open a connection at the end (think heavy app generating logs and dumping them somewhere else). So this won't work in general.

I'm actually would be totally OK if we stopped using envoy for readiness and prometheus ports. I don't like the complexity incurred by having to deal with those listeners. It's complexity for complexity sake.

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | f631a8bdc9527df6245ea11f0a731d246628d257 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435487266393821184) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 345893451d34b6715c10617931cbe9956e522b74 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/release-notes_istio/1435167448427925504) | `/test release-notes_istio`
unit-tests_istio | 2bd92eda30db8d1b9390283cfd4a78a3d9f1f48a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/35043/unit-tests_istio/1435488881293135872) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test unit-tests_istio

ramaraochavali commented:
@howardjohn @kyessenov Some how this branch is messedup. Closing this and opening a new PR https://github.com/istio/istio/pull/35059.

@howardjohn addressed all your comments. If we are good with the approach I will add tests.

howardjohn commented:
> The cleanest solution is to wait for other containers to complete IMHO (and dealing with their organic restarts).

I agree but don't know how to do this. Any ideas?
